### PR TITLE
Add base project

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

Add base project to work on.

## Proposal

- [x] add `fun-mooc` Django project
- [x] add base Docker environment
- [x] add development Docker image
- [x] add Makefile helpers
- [x] add base documentation

### CODE REVIEW

File: docker/files/usr/local/bin/entrypoint [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+# The following environment variables may be passed to the container to
+# customize running user account:
+#
+#   * USER_NAME: container user name (default: default)
+#   * HOME     : container user home directory (default: none)
+#
+# To pass environment variables, you can either use the -e option of the docker
+# run command:
+#
+#     docker run --rm -e USER_NAME=foo -e HOME='/home/foo' richie:latest python sandbox/manage.py migrate
```
Comment (jmaupetit) [inline docker/files/usr/local/bin/entrypoint]: Should not be Richie.

Comment (jmaupetit) [reply docker/files/usr/local/bin/entrypoint]: Fix management script path.

File: docker/files/usr/local/bin/entrypoint [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#!/bin/sh
```
Comment (jmaupetit) [inline docker/files/usr/local/bin/entrypoint]: Should be bash.

File: docker/files/usr/local/bin/entrypoint [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+#
+#     # env.d/production
+#     USER_NAME=foo
+#     HOME=/home/foo
+#
+#     docker run --rm --env-file env.d/production richie:latest python sandbox/manage.py migrate
+#
+
+echo "🐳(entrypoint) creating user running in the container..."
+if ! whoami &> /dev/null; then
```
Comment (jmaupetit) [inline docker/files/usr/local/bin/entrypoint]: Strange bash redirection.

File: env.d/development [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+# Django
+DJANGO_SETTINGS_MODULE=funmooc.settings
+DJANGO_CONFIGURATION=Development
+DJANGO_SECRET_KEY=ThisIsAnExampleKeyForDevPurposeOnly
+
+# Elastic search
+ES_CLIENT=elasticsearch
+
+# PostgreSQL db container configuration
+POSTGRES_DB=richie
```
Comment (jmaupetit) [inline env.d/development]: fun-mooc !

File: env.d/development [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+POSTGRES_USER=fun
+POSTGRES_PASSWORD=pass
+
+# App database configuration
+DB_HOST=db
+DB_NAME=richie
```
Comment (jmaupetit) [inline env.d/development]: fun-mooc !

File: src/backend/funmooc/wsgi.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+For more information on this file, see
+https://docs.djangoproject.com/en/1.8/howto/deployment/wsgi/
+"""
+
+import os
+
+from configurations.wsgi import get_wsgi_application
+
+os.environ.setdefault("DJANGO_SETTINGS_MODULE", "settings")
```
Comment (jmaupetit) [inline src/backend/funmooc/wsgi.py]:
```suggestion
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "funmooc.settings")
```

File: src/backend/funmooc/settings.py [hunk 7]
```diff
@@ -0,0 +1,640 @@
...
+
+    # Database
+    DATABASES = {
+        "default": {
+            "ENGINE": values.Value(
+                "django.db.backends.postgresql_psycopg2",
+                environ_name="DB_ENGINE",
+                environ_prefix=None,
+            ),
+            "NAME": values.Value("richie", environ_name="DB_NAME", environ_prefix=None),
```
Comment (jmaupetit) [inline src/backend/funmooc/settings.py]: fun-mooc!

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+# -- Rules
+default: help
+
+bootstrap: data/media/.keep data/static/.keep build run migrate  ## install development dependencies
+.PHONY: bootstrap
+
+# == Docker
+build: ## build the app container
+	@$(COMPOSE) build --build-arg UID=$(UID) app
```
Comment (lunika) [inline Makefile]: are you angry with your tabulation key ?

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+search-index:  ## (re)generate the Elasticsearch index
+	@$(MANAGE) bootstrap_elasticsearch
+.PHONY: search-index
+
+superuser: ## create a DjangoCMS superuser
+	@$(MANAGE) createsuperuser
+.PHONY: superuser
+
+# == Misc
+clean: ## restore repository state as it was freshly cloned
```
Comment (lunika) [inline Makefile]: `.PHONY` is missing for this task please.

Comment (jmaupetit) [reply Makefile]: Well spotted!

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+nosetests.xml
+coverage.xml
+
+# Vagrant
+.vagrant
+
+# PyCharm IDE
+.idea/
+
+# Mr Developer
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply .gitignore]: I do not fully agree with this: this is useful for contributors that set `.gitignore` rules on a per-project basis. This is why I took GitHub template for python projects: https://github.com/github/gitignore/blob/master/Python.gitignore

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+# Run django development server (wrapped by dockerize to ensure the db is ready
+# to accept connections before running the development server)
+CMD dockerize -wait tcp://db:5432 -timeout 60s \
+    python manage.py runserver 0.0.0.0:8000
+
+# ---- Production image ----
+FROM core as production
+
+# The default command runs gunicorn WSGI server in the sandbox
+CMD gunicorn -c /usr/local/etc/gunicorn/funmooc.py funmooc.wsgi:application
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply Dockerfile]: In fact, the latest stage (default build) should be the production image, this is why I decided to split things this way.

File: .gitignore (hunk 10 above)
[the remark this answers is not in the document]
Comment (lunika) [reply .gitignore]: :+1: for removing IDE specific files

[the remark this answers is not in the document]
Comment (lunika) [reply .gitignore]: I dont see them in this template: https://github.com/github/gitignore/blob/master/Python.gitignore

[the remark this answers is not in the document]
Comment (lunika) [reply .gitignore]: As we say in france :

:musical_note:  Et ils sont où ? Et ils sont où les IDE ? :musical_note:

[the remark this answers is not in the document]
Comment (jmaupetit) [reply .gitignore]: You. Are. Right. I was pretty sure that the IDE section came from this original file. My apologizes. I'll fix this.

File: .gitignore [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+/*.sublime-workspace
+
+# TextMate 2
+.tm_properties
+
+# VS Code
+.vscode
+
+# Mac Os
+.DS_Store
```
Comment (lunika) [inline .gitignore]: This should also be removed, this is not our responsibility to add it. And it is not in the Python.gitignore file :-)

Comment (jmaupetit) [reply .gitignore]: Done.

File: Dockerfile (hunk 11 above)
[the remark this answers is not in the document]
Comment (lunika) [reply Dockerfile]: I agree with @jmaupetit

[the remark this answers is not in the document]
Comment (sampaccoud) [reply Dockerfile]: IC

**Comment by jmaupetit** [comment]
@lunika / @sampaccoud I've fixed all your comments. Are good to go?

**Pull request merged**